### PR TITLE
Pulp & Patoh Integration

### PR DESCRIPTION
**Pull request by ardasener**
This PR includes the following contributions:
- `PulpPartition` class wrapping the partitioner available here: https://github.com/HPCGraphAnalysis/PuLP
- `PatohPartition` class wrapping the partitioner available here: https://faculty.cc.gatech.edu/~umit/software.html
- Tests for the aforementioned classes and the previously introduced `MetisPartition`.
- Update to the optional dependencies page in docs with a warning regarding static libraries.
- Standardization of `@param` style doc comments
- Update to the Testing action on GitHub to test with PATOH and PULP as well.
- Default template types are replaced with their fixed-width counterparts (https://en.cppreference.com/w/cpp/types/integer). 
  - By default, 32 and 64-bit signed and unsigned integers are supported for IDType and NNZType

As of now, the following work is still needed for the partitioning side of this library to be complete:
- A class to apply partitionings to formats (Not straightforward due to the many different ways one can apply a partitioning).
- A `PartitionBase` class similar to `ReorderBase`.
- Passing weights to the partitioners is currently not supported. This should be added soon.
- Some partitioners are only designed for graphs (square matrices). I believe all the hypergraph partitioners should work with matrices in general but I am not 100% sure. Regardless this should be indicated to the user perhaps by renaming the classes to `XXXGraphPartitioner`. And assertions can be added as well.

### CODE REVIEW

File: src/sparsebase/preprocess/preprocess.cc [hunk 1]
```diff
@@ -1634,6 +1634,166 @@ IDType *MetisReorder<IDType, NNZType, ValueType>::GetReorderCSR(
...
+
+  if constexpr (std::is_signed_v<IDType> && std::is_signed_v<NNZType> &&
+                sizeof(IDType) == sizeof(int) &&
+                sizeof(NNZType) == sizeof(long)) {
+    pulp_graph_t graph;
+    graph.n = n;
+    graph.m = m;
+    graph.out_array = csr->get_col();
+    graph.out_degree_list = csr->get_row_ptr();
+    graph.vertex_weights = nullptr;
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:1692]: maybe we should consider supporting weights in formats

File: src/sparsebase/preprocess/preprocess.cc [hunk 2]
```diff
@@ -1634,6 +1634,166 @@ IDType *MetisReorder<IDType, NNZType, ValueType>::GetReorderCSR(
...
+  this->params_ =
+      std::unique_ptr<PatohPartitionParams>(new PatohPartitionParams(params));
+}
+
+
+template <typename IDType, typename NNZType, typename ValueType>
+IDType *PatohPartition<IDType, NNZType, ValueType>::PartitionCSR(
+    std::vector<format::Format *> formats, PreprocessParams *params){
+
+  if(typeid(IDType) != typeid(int) || typeid(NNZType) != typeid(int)){
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:1731]: constexpr like above?

**Review by Atahanak — APPROVED** [review]
left minor comments.
Seems good.

File: src/sparsebase/preprocess/preprocess.cc [hunk 3]
```diff
@@ -1634,6 +1634,166 @@ IDType *MetisReorder<IDType, NNZType, ValueType>::GetReorderCSR(
...
+  con.do_lp_init = pparams->do_lp_init;
+  con.do_bfs_init = pparams->do_bfs_init;
+  con.do_repart = pparams->do_repart;
+  con.do_edge_balance = pparams->do_edge_balance;
+  con.do_maxcut_balance = pparams->do_maxcut_balance;
+
+  int np = pparams->num_partitions;
+  IDType* partition = new IDType[n];
+
+  if constexpr (std::is_signed_v<IDType> && std::is_signed_v<NNZType> &&
```
Comment (Atahanak) [inline src/sparsebase/preprocess/preprocess.cc:1684]: try converting first and if that fails raise an exception could be better.

File: src/sparsebase/preprocess/preprocess.cc (hunk 2 above)
Comment (Atahanak) [reply src/sparsebase/preprocess/preprocess.cc:1731]: same here converter can be used.

**Review by Atahanak — CHANGES_REQUESTED** [review]
minor changes needed.

File: src/sparsebase/preprocess/preprocess.cc (hunk 1 above)
Comment (ardasener) [reply src/sparsebase/preprocess/preprocess.cc:1692]: It is not trivial, they all support different types which is why I avoided doing so for now.

**Comment by AmroAlJundi** [comment]
Btw, don't forget to add the `+= USE_PULP` line to `docs/conf.py` and to add pages for the new partitionings to the `available.md` page.

**Review by Atahanak — APPROVED** [review]

**Review by Atahanak — APPROVED** [review]

**Pull request merged**